wgsl: scalar-matrix multiplication not correctly rounding operations

Chromium/Dawn is reporting errors that look like this:
```
AssertionError:   - EXPECTATION FAILED: (mat2x2<f32>(-3.4028234663852886e+38 (0xff7fffff), 0.0 (0x80000000 subnormal), 0.0 (0x80000000 subnormal), 0.0 (0x00000000 subnormal)), -1.1754942106924411e-38 (0x807fffff subnormal))
        returned: mat2x2<f32>(-1.1754942106924411e-38 (0x807fffff subnormal), -1.1754942106924411e-38 (0x807fffff subnormal), -1.1754942106924411e-38 (0x807fffff subnormal), -1.1754942106924411e-38 (0x807fffff subnormal))
        expected: [[[{ 'f32', [0.0 (0x80000000 subnormal),3.999999523162842 (0x407ffffe)] }],[{ 'f32', [0.0 (0x80000000 subnormal)] }]],[[{ 'f32', [0.0 (0x80000000 subnormal)] }],[{ 'f32', [0.0 (0x80000000 subnormal)] }]]]
    
    (mat2x2<f32>(-3.4028234663852886e+38 (0xff7fffff), 0.0 (0x80000000 subnormal), 0.0 (0x80000000 subnormal), 0.0 (0x00000000 subnormal)), -1.401298464324817e-45 (0x80000001 subnormal))
        returned: mat2x2<f32>(-1.401298464324817e-45 (0x80000001 subnormal), -1.401298464324817e-45 (0x80000001 subnormal), -1.401298464324817e-45 (0x80000001 subnormal), -1.401298464324817e-45 (0x80000001 subnormal))
        expected: [[[{ 'f32', [0.0 (0x80000000 subnormal),4.7683712978141557e-7 (0x34ffffff)] }],[{ 'f32', [0.0 (0x80000000 subnormal)] }]],[[{ 'f32', [0.0 (0x80000000 subnormal)] }],[{ 'f32', [0.0 (0x80000000 subnormal)] }]]]
    
    (mat2x2<f32>(-3.4028234663852886e+38 (0xff7fffff), 0.0 (0x80000000 subnormal), 0.0 (0x80000000 subnormal), 0.0 (0x00000000 subnormal)), 1.401298464324817e-45 (0x00000001 subnormal))
        returned: mat2x2<f32>(1.401298464324817e-45 (0x00000001 subnormal), 1.401298464324817e-45 (0x00000001 subnormal), 1.401298464324817e-45 (0x00000001 subnormal), 1.401298464324817e-45 (0x00000001 subnormal))
        expected: [[[{ 'f32', [-4.7683712978141557e-7 (0xb4ffffff),0.0 (0x80000000 subnormal)] }],[{ 'f32', [0.0 (0x80000000 subnormal)] }]],[[{ 'f32', [0.0 (0x80000000 subnormal)] }],[{ 'f32', [0.0 (0x00000000 subnormal)] }]]]
```

The critical part here is that we have a mat2x2 * a subnormal scalar. Since that scalar can be FTZ, one of the acceptable answers should be the zero matrix, which in turn should accept elements that are subnormal, since they can be FTZ.

The rejected matrix is all subnormal values.

Looking at the code, it looks like multiplicationMatrixScalarInterval is reaching directly into the multiplicationInterval impl instead of calling multiplicationInterval, which is like the source of this bug, since correctlyRounded will be applied by the scaffolding code called by multiplicationInterval to invoke the impl

--- a/src/unittests/floating_point.spec.ts
+++ b/src/unittests/floating_point.spec.ts
@@ -7574,135 +7574,171 @@ interface MatrixScalarToMatrixCase {
   expected: (number | IntervalBounds)[][];
 }
 
+const kMultiplicationMatrixScalarIntervalCases = {
+  f32: [
+    // From https://github.com/gpuweb/cts/issues/3044
+    {
+      matrix: [
+        [kValue.f32.negative.min, 0],
+        [0, 0],
+      ],
+      scalar: kValue.f32.negative.subnormal.min,
+      expected: [
+        [[0, reinterpretU32AsF32(0x407ffffe)], 0], // [[0, 3.9999995...], 0],
+        [0, 0],
+      ],
+    },
+  ] as MatrixScalarToMatrixCase[],
+  f16: [
+    // From https://github.com/gpuweb/cts/issues/3044
+    {
+      matrix: [
+        [kValue.f16.negative.min, 0],
+        [0, 0],
+      ],
+      scalar: kValue.f16.negative.subnormal.min,
+      expected: [
+        [[0, reinterpretU16AsF16(0x43fe)], 0], // [[0, 3.99609375], 0]
+        [0, 0],
+      ],
+    },
+  ] as MatrixScalarToMatrixCase[],
+} as const;
+
 g.test('multiplicationMatrixScalarInterval')
   .params(u =>
     u
       .combine('trait', ['f32', 'f16'] as const)
       .beginSubcases()
-      .combineWithParams<MatrixScalarToMatrixCase>([
-        // Only testing that different shapes of matrices are handled correctly
-        // here, to reduce test duplication.
-        // multiplicationMatrixScalarInterval uses MultiplicationIntervalOp for calculating intervals,
-        // so the testing for multiplcationInterval covers the actual interval
+      .expandWithParams<MatrixScalarToMatrixCase>(p => {
+        // Primarily testing that different shapes of matrices are handled correctly
+        // here, to reduce test duplication. Additional testing for edge case
+        // discovered in https://github.com/gpuweb/cts/issues/3044.
+        //
+        // multiplicationMatrixScalarInterval uses for calculating intervals,
+        // so the testing for multiplicationInterval covers the actual interval
         // calculations.
-        {
-          matrix: [
-            [1, 2],
-            [3, 4],
-          ],
-          scalar: 10,
-          expected: [
-            [10, 20],
-            [30, 40],
-          ],
-        },
-        {
-          matrix: [
-            [1, 2],
-            [3, 4],
-            [5, 6],
-          ],
-          scalar: 10,
-          expected: [
-            [10, 20],
-            [30, 40],
-            [50, 60],
-          ],
-        },
-        {
-          matrix: [
-            [1, 2],
-            [3, 4],
-            [5, 6],
-            [7, 8],
-          ],
-          scalar: 10,
-          expected: [
-            [10, 20],
-            [30, 40],
-            [50, 60],
-            [70, 80],
-          ],
-        },
-        {
-          matrix: [
-            [1, 2, 3],
-            [4, 5, 6],
-          ],
-          scalar: 10,
-          expected: [
-            [10, 20, 30],
-            [40, 50, 60],
-          ],
-        },
-        {
-          matrix: [
-            [1, 2, 3],
-            [4, 5, 6],
-            [7, 8, 9],
-          ],
-          scalar: 10,
-          expected: [
-            [10, 20, 30],
-            [40, 50, 60],
-            [70, 80, 90],
-          ],
-        },
-        {
-          matrix: [
-            [1, 2, 3],
-            [4, 5, 6],
-            [7, 8, 9],
-            [10, 11, 12],
-          ],
-          scalar: 10,
-          expected: [
-            [10, 20, 30],
-            [40, 50, 60],
-            [70, 80, 90],
-            [100, 110, 120],
-          ],
-        },
-        {
-          matrix: [
-            [1, 2, 3, 4],
-            [5, 6, 7, 8],
-          ],
-          scalar: 10,
-          expected: [
-            [10, 20, 30, 40],
-            [50, 60, 70, 80],
-          ],
-        },
-        {
-          matrix: [
-            [1, 2, 3, 4],
-            [5, 6, 7, 8],
-            [9, 10, 11, 12],
-          ],
-          scalar: 10,
-          expected: [
-            [10, 20, 30, 40],
-            [50, 60, 70, 80],
-            [90, 100, 110, 120],
-          ],
-        },
-        {
-          matrix: [
-            [1, 2, 3, 4],
-            [5, 6, 7, 8],
-            [9, 10, 11, 12],
-            [13, 14, 15, 16],
-          ],
-          scalar: 10,
-          expected: [
-            [10, 20, 30, 40],
-            [50, 60, 70, 80],
-            [90, 100, 110, 120],
-            [130, 140, 150, 160],
-          ],
-        },
-      ])
+        return [
+          {
+            matrix: [
+              [1, 2],
+              [3, 4],
+            ],
+            scalar: 10,
+            expected: [
+              [10, 20],
+              [30, 40],
+            ],
+          },
+          {
+            matrix: [
+              [1, 2],
+              [3, 4],
+              [5, 6],
+            ],
+            scalar: 10,
+            expected: [
+              [10, 20],
+              [30, 40],
+              [50, 60],
+            ],
+          },
+          {
+            matrix: [
+              [1, 2],
+              [3, 4],
+              [5, 6],
+              [7, 8],
+            ],
+            scalar: 10,
+            expected: [
+              [10, 20],
+              [30, 40],
+              [50, 60],
+              [70, 80],
+            ],
+          },
+          {
+            matrix: [
+              [1, 2, 3],
+              [4, 5, 6],
+            ],
+            scalar: 10,
+            expected: [
+              [10, 20, 30],
+              [40, 50, 60],
+            ],
+          },
+          {
+            matrix: [
+              [1, 2, 3],
+              [4, 5, 6],
+              [7, 8, 9],
+            ],
+            scalar: 10,
+            expected: [
+              [10, 20, 30],
+              [40, 50, 60],
+              [70, 80, 90],
+            ],
+          },
+          {
+            matrix: [
+              [1, 2, 3],
+              [4, 5, 6],
+              [7, 8, 9],
+              [10, 11, 12],
+            ],
+            scalar: 10,
+            expected: [
+              [10, 20, 30],
+              [40, 50, 60],
+              [70, 80, 90],
+              [100, 110, 120],
+            ],
+          },
+          {
+            matrix: [
+              [1, 2, 3, 4],
+              [5, 6, 7, 8],
+            ],
+            scalar: 10,
+            expected: [
+              [10, 20, 30, 40],
+              [50, 60, 70, 80],
+            ],
+          },
+          {
+            matrix: [
+              [1, 2, 3, 4],
+              [5, 6, 7, 8],
+              [9, 10, 11, 12],
+            ],
+            scalar: 10,
+            expected: [
+              [10, 20, 30, 40],
+              [50, 60, 70, 80],
+              [90, 100, 110, 120],
+            ],
+          },
+          {
+            matrix: [
+              [1, 2, 3, 4],
+              [5, 6, 7, 8],
+              [9, 10, 11, 12],
+              [13, 14, 15, 16],
+            ],
+            scalar: 10,
+            expected: [
+              [10, 20, 30, 40],
+              [50, 60, 70, 80],
+              [90, 100, 110, 120],
+              [130, 140, 150, 160],
+            ],
+          },
+          ...kMultiplicationMatrixScalarIntervalCases[p.trait],
+        ];
+      })
   )
   .fn(t => {
     const matrix = t.params.matrix;

--- a/src/webgpu/util/floating_point.ts
+++ b/src/webgpu/util/floating_point.ts
@@ -3748,7 +3748,7 @@ export abstract class FPTraits {
     const rows = mat[0].length;
     return this.toMatrix(
       unflatten2DArray(
-        flatten2DArray(mat).map(e => this.MultiplicationIntervalOp.impl(e, scalar)),
+        flatten2DArray(mat).map(e => this.multiplicationInterval(e, scalar)),
         cols,
         rows
       )

